src/field/arena.ts

```diff
@@ -357,22 +350,33 @@ export class Arena {
    * @param terrain - {@linkcode TerrainType} or array of {@linkcode TerrainType} to check against
    * @returns `true` if the arena is of the specified terrain, `false` otherwise
    */
-  public hasTerrain(terrain: TerrainType | TerrainType[]): boolean {
-    return coerceArray(terrain).includes(this.terrainType);
+  public hasTerrain(...terrain: readonly [TerrainType, ...TerrainType[]]): boolean {
+    return terrain.includes(this.terrainType);
+  }
+
+  /** Sets terrain to the override specified in `overrides.ts` */
+  private tryOverrideTerrain(): boolean {
+    const newTerrain = activeOverrides.TERRAIN_OVERRIDE;
+    if (newTerrain === TerrainType.NONE) {
+      return false;
+    }
+
+    this.terrain = new Terrain(newTerrain, 0);
+    globalScene.phaseManager.createAndUnshiftPhase("CommonAnimPhase", getTerrainAnim(newTerrain));
+    globalScene.phaseManager.createAndUnshiftPhase("MessagePhase", getTerrainStartMessage(newTerrain));
+
+    return true;
   }
 
   /**
-   * Sets terrain to the override specified in overrides.ts
-   * @param terrain new {@linkcode TerrainType} to set
-   * @returns true to force trySetTerrain to return true
+   * Check if it's possible to change the terrain
+   * @param terrain - The {@linkcode TerrainType} to set
+   * @returns Whether the terrain can be changed
    */
-  private tryOverrideTerrain(terrain: TerrainType): boolean {
-    this.terrain = new Terrain(terrain, 0);
-    globalScene.phaseManager.createAndUnshiftPhase(
-      "CommonAnimPhase",
-      (CommonAnim.MISTY_TERRAIN + (terrain - 1)) as CommonAnim,
-    );
-    globalScene.phaseManager.createAndUnshiftPhase("MessagePhase", getTerrainStartMessage(terrain) ?? "");
+  public canSetTerrain(terrain: TerrainType): boolean {
+    if (terrain === this.terrainType) {
+      return false;
+    }
     return true;
```

Comment:
```suggestion
    return terrain !== this.terrainType;
```

Also consider renaming `terrain` to `terrainType`